### PR TITLE
Request Response charges link

### PR DESCRIPTION
**Pull request by Korgath**
## Flow
image

## Still to do
- Handle response on how sending out messages went
- <del>Implement Outbox steps <del>

## To discuss:
- <del>Come to a decision on whether CorrelationId should be added to the Metering Point Table, or should be sent with the message dequeued event (or third a third option)<del>
- Error handling if response from Charges contain an error

## References

* #0000

### CODE REVIEW

File: source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/ChargesResponseReceiver.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        }
+
+        [Function("ChargesResponseReceiver")]
+        public async Task RunAsync([ServiceBusTrigger("%CHARGES_RESPONSE_QUEUE%", Connection = "SHARED_INTEGRATION_EVENT_SERVICE_BUS_SENDER_CONNECTION_STRING")] byte[] data, FunctionContext context)
+        {
+            if (context == null) throw new ArgumentNullException(nameof(context));
+            _logger.LogInformation($"Received an response from Charges.");
+
+            var reader = new DefaultChargeLinkReplyReader(HandleSuccessAsync, HandleFailureAsync);
+            await reader.ReadAsync(data, RequestStatus.Succeeded).ConfigureAwait(false);
```
Comment (Korgath) [inline source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/ChargesResponseReceiver.cs:51]: This is awaiting some package updates - hence it is always status Succeeded

File: source/Energinet.DataHub.MeteringPoints.ApplyDBMigrationsApp/Energinet.DataHub.MeteringPoints.ApplyDBMigrationsApp.csproj [hunk 2]
```diff
@@ -55,6 +55,8 @@ limitations under the License.
     <EmbeddedResource Include="Scripts\Model\202109271100 Add IsOfficialAddress.sql" />
     <EmbeddedResource Include="Scripts\Model\202109271400 Add GeoInfoReference.sql" />
     <EmbeddedResource Include="Scripts\Model\202109270930 Add MessageHubMessages table.sql" />
+    <None Remove="Scripts\Model\202110080920 Alter Correlation in QueuedInternalCommands.sql" />
```
Comment (thbaan) [inline source/Energinet.DataHub.MeteringPoints.ApplyDBMigrationsApp/Energinet.DataHub.MeteringPoints.ApplyDBMigrationsApp.csproj:58]: Can this line be removed?

File: source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/ChargesResponseReceiver.cs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+using Energinet.DataHub.Charges.Libraries.DefaultChargeLink;
+using Energinet.DataHub.Charges.Libraries.Enums;
+using Energinet.DataHub.Charges.Libraries.Models;
+using Energinet.DataHub.MeteringPoints.Application.Integrations.ChargeLinks.Create;
+using Energinet.DataHub.MeteringPoints.Infrastructure.Correlation;
+using MediatR;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Extensions.Logging;
+
+namespace Energinet.DataHub.MeteringPoints.EntryPoints.Processing
```
Comment (thbaan) [inline source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/ChargesResponseReceiver.cs:26]: Maybe we should introduce a folder named "Functions" where this, IntegrationEventReceiver and QueueSubscriber can be placed in?

File: source/Energinet.DataHub.MeteringPoints.Infrastructure/InternalCommands/Protobuf/Mappers/CreateDefaultChargeLinksMessages/CreateDefaultChargeLinksMessagesOutboundMapper.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        protected override IMessage Convert(Application.Integrations.ChargeLinks.Messages.CreateDefaultChargeLinksMessages obj)
+        {
+            if (obj == null) throw new ArgumentNullException(nameof(obj));
+
+            return new MeteringPointEnvelope
+            {
+                CreateDefaultChargeLinksMessages = new Contracts.CreateDefaultChargeLinksMessages()
+                {
+                    CorrelationId = obj.CorrelationId, GsrnNumber = obj.GsrnNumber,
```
Comment (thbaan) [inline source/Energinet.DataHub.MeteringPoints.Infrastructure/InternalCommands/Protobuf/Mappers/CreateDefaultChargeLinksMessages/CreateDefaultChargeLinksMessagesOutboundMapper.cs:32]: Should be located on separate lines

File: source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/EventMetadata.cs [hunk 5]
```diff
@@ -15,7 +15,7 @@
 namespace Energinet.DataHub.MeteringPoints.EntryPoints.Processing
 {
         public class EventMetadata
```
Comment (cmgrd) [inline source/Energinet.DataHub.MeteringPoints.EntryPoints.Processing/EventMetadata.cs:17]: nit: indentation

**Review by cmgrd — DISMISSED** [review]

**Review by thbaan — APPROVED** [review]

File: source/Energinet.DataHub.MeteringPoints.EntryPoints.Outbox/Program.cs [hunk 6]
```diff
@@ -97,6 +101,20 @@ protected override void ConfigureContainer(Container container)
                 () => new ServiceBusClient(connectionString),
                 Lifestyle.Singleton);
 
+            // SB for communicating with Charges
+            container.Register<DefaultChargeLinkClient>(
+                () => new DefaultChargeLinkClient(
+                    container.GetInstance<ServiceBusClient>(),
+                    new ServiceBusRequestSenderFactory(),
+                    Environment.GetEnvironmentVariable("CHARGES_DEFAULT_LINK_RESPONSE_QUEUE") ?? throw new InvalidOperationException()),
```
Comment (cmgrd) [inline source/Energinet.DataHub.MeteringPoints.EntryPoints.Outbox/Program.cs:109]: Are these queues available via the same service bus namespace as `SHARED_INTEGRATION_EVENT_SERVICE_BUS_SENDER_CONNECTION_STRING`?

**Review by thbaan — APPROVED** [review]

**Review by cmgrd — APPROVED** [review]

**Pull request merged**